File: clojure/problem-3.clj
; The prime factors of 13195 are 5, 7, 13 and 29.
; What is the largest prime factor of the number 600851475143 ?

; http://www.wikihow.com/Check-if-a-Number-Is-Prime
; The Murderous Method
; ((n²)+17)/12 = x.5


; 1st attempt...horribly inefficient for large numbers

; (def limit 600851475143)
; (defn factor? [n d] 
;   (if (zero? d)
;     false
;     (zero? (mod n d))))
; (defn factors [n]
;   (filter #(factor? n %) (range 1 (+ n 1))))
; (defn prime? [n]
;   (= 2 (count (factors n))))

; (take-last 1 (filter #(prime? %) (range limit)))


; 2nd attempt...much better computationally and I admit I got a few ideas from others


(def limit 600851475143)

(defn factor? [n d] 
	(if (zero? d)
		false
		(zero? (mod n d))))

(defn lpf [n]
	(let [q (long (clojure.Math/sqrt n))]			    ; quadratic limit
    	(loop [n n
        	   d 2]
      		(cond
       			(> d q) n                           ; num itself must be prime
       			(= n d) n                           ; n must be prime
       			(factor? n d) (recur (/ n d) d)     ; factor d out of n
       			true          (recur n (inc d)))))) ; else try again with bigger d

(lpf limit)

File: go/problem3.go
// The prime factors of 13195 are 5, 7, 13 and 29.

// What is the largest prime factor of the number 600851475143 ?

package main

import (
  "fmt"
  "math"
)

func is_factor(n, d int64) bool {
  if(d == 0) {
    return false
  } else {
    return n % d == 0
  }
}

func main() {
  var limit int64 = 600851475143
  var q, n, d int64 = int64(math.Sqrt(float64(limit))), limit, 2

  for  ; n >= d ; {
    if (d > q) || (n == d) {
      fmt.Println(n)
      break
    } else if is_factor(n, d) {
      n = n / d
    } else {
      d++
    }
  }
}